### PR TITLE
Implementation of ATKD

### PR DESCRIPTION
**Pull request by akshaykvnit**
Adding implementation of [Attention Transfer KD (ATKD)](https://arxiv.org/abs/1612.03928) which is an [ICLR '17 paper](https://openreview.net/forum?id=Sks9_ajex). Some inspiration for the source code is from their [official implementation](https://github.com/szagoruyko/attention-transfer).

Please go through the code (in case there are any obvious mistakes).

I ran one experiment with ResNet10 full data and the result is 92% validation accuracy (comparable to 92.2% of simultaneous KD for same settings).

### CODE REVIEW

**Comment by akshaykvnit** [comment]
Loss Function of ATKD
The authors mention that they use beta = 1000 / (batch size * number of elements in attention map) (around 0.1). They also say that they decay this beta parameter when using it along with KD.
Now, we have 2 choices:

1. Use beta = 1 because our implementation doesn't rely on such weighting.
2. Use beta as in the paper. However, they don't mention how they actually decay it and their code is not exactly readable (but if you can go through and find, then it would be good).

Please advise on what to do? As far as implementation is concerned, both are equally straightforward.

**Comment by SharathRaparthy** [comment]
> Loss Function of ATKD
> The authors mention that they use beta = 1000 / (batch size * number of elements in attention map) (around 0.1). They also say that they decay this beta parameter when using it along with KD.
> Now, we have 2 choices:
> 
> 1. Use beta = 1 because our implementation doesn't rely on such weighting.
> 2. Use beta as in the paper. However, they don't mention how they actually decay it and their code is not exactly readable (but if you can go through and find, then it would be good).
> 
> Please advise on what to do? As far as implementation is concerned, both are equally straightforward.

Implement #1 as of now and let's see how the experiments go. In the meanwhile, you can look into the decaying. Are they using any schedulers?

**Comment by akshaykvnit** [comment]
@navidpanchi @SharathRaparthy 
Some more information of their experiments:
1. Use SGD with weight decay (1e-4) (L2 regularization) (we use Adam without weight decay)
2. Use learning rate scheduling in steps. Multiply LR by 0.1 at 30, 60, 90 epochs. (we don't use any LR scheduling in image classification experiments).
3. No information about decaying the beta parameter. They mention in their README that they plan to add it, but last commit was in July 2018. There is also an [open issue](https://github.com/szagoruyko/attention-transfer/issues/33) regarding this without reply.
